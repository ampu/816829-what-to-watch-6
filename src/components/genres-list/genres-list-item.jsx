import React, {useCallback} from 'react';
import {Link} from 'react-router-dom';
import PropTypes from 'prop-types';
import getClassName from 'classnames';

import {MainPath} from '../../constants/paths';

const GenresListItem = ({genre, onGenreChange, isActive = false}) => {

  const classMap = {
    [`catalog__genres-item`]: true,
    [`catalog__genres-item--active`]: isActive,
  };

  const handleLinkClick = useCallback((evt) => {
    evt.preventDefault();
    onGenreChange(genre);
  }, [onGenreChange, genre]);

  return (
    <li className={getClassName(classMap)}>
      <Link className="catalog__genres-link" to={MainPath.INDEX} onClick={handleLinkClick}>{genre}</Link>
    </li>
  );
};

GenresListItem.propTypes = {
  onGenreChange: PropTypes.func.isRequired,
  genre: PropTypes.string.isRequired,
  isActive: PropTypes.bool,
};

export default GenresListItem;
