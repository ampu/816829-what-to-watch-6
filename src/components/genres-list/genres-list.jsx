import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';

import GenresListItem from './genres-list-item';
import {setGenre} from '../../store/actions/genre-actions';
import {selectGenres} from '../../store/selectors';

const GenresList = ({genres = [], activeGenre = ``, onGenreChange}) => {
  return (
    <ul className="catalog__genres-list">
      {genres.map((genre) => (
        <GenresListItem key={genre}
          genre={genre}
          onGenreChange={onGenreChange}
          isActive={genre === activeGenre}/>
      ))}
    </ul>
  );
};

GenresList.propTypes = {
  genres: PropTypes.arrayOf(PropTypes.string),
  activeGenre: PropTypes.string,
  onGenreChange: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => ({
  genres: selectGenres(state),
  activeGenre: state.genre,
});

const mapDispatchToProps = (dispatch) => ({
  onGenreChange(genre) {
    dispatch(setGenre(genre));
  },
});

export {GenresList};
export default connect(mapStateToProps, mapDispatchToProps)(GenresList);
