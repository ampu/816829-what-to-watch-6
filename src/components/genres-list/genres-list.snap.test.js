import React from 'react';
import {Router} from 'react-router-dom';
import {createMemoryHistory} from 'history';

import {render} from '@testing-library/react';

import {GenresList} from './genres-list';

describe(`GenresList component`, () => {
  it(`Should render according to the snapshot with no genres`, () => {
    const {container} = render(<GenresList onGenreChange={jest.fn()} genres={[]}/>);
    expect(container).toMatchSnapshot();
  });

  it(`Should render according to the snapshot with genres provided`, () => {
    const genres = [`Main`, `Auxiliary`];
    const {container} = render((
      <Router history={createMemoryHistory()}>
        <GenresList onGenreChange={jest.fn()} genres={genres} activeGenre="Auxiliary"/>;
      </Router>
    ));
    expect(container).toMatchSnapshot();
  });
});
