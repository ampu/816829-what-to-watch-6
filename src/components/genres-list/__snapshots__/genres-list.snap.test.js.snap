// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenresList component Should render according to the snapshot with genres provided 1`] = `
<div>
  <ul
    class="catalog__genres-list"
  >
    <li
      class="catalog__genres-item"
    >
      <a
        class="catalog__genres-link"
        href="/"
      >
        Main
      </a>
    </li>
    <li
      class="catalog__genres-item catalog__genres-item--active"
    >
      <a
        class="catalog__genres-link"
        href="/"
      >
        Auxiliary
      </a>
    </li>
  </ul>
  ;
</div>
`;

exports[`GenresList component Should render according to the snapshot with no genres 1`] = `
<div>
  <ul
    class="catalog__genres-list"
  />
</div>
`;
